native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt

```diff
@@ -17,7 +17,33 @@ class DatabaseDriver(context: Context, private val configuration: Configuration)
         // class RemoveLocal(val key: String) : Operation()
     }
 
-    private val database: Database by lazy { Database(configuration.name, context) }
+    class SchemaNeededError : Exception()
+    data class MigrationNeededError(val databaseVersion: SchemaVersion) : Exception()
+
+    constructor(context: Context, dbName: String, schemaVersion: SchemaVersion)
+            : this(context, dbName) {
+        val compatibility = isCompatible(schemaVersion)
+        when (compatibility) {
+            is SchemaCompatibility.NeedsSetup -> throw SchemaNeededError()
+            is SchemaCompatibility.NeedsMigration ->
+                throw MigrationNeededError(compatibility.fromVersion)
+        }
+    }
+
+    constructor(context: Context, dbName: String, schema: Schema) : this(context, dbName) {
+        setUpDatabase(schema)
+    }
+
+    constructor(context: Context, dbName: String, migrations: MigrationSet) :
+            this(context, dbName) {
+        migrate(migrations)
+    }
+
+    constructor(context: Context, dbName: String) {
```

Comment:
I think that this should be a primary constructor and I am wondering if maybe factory pattern wouldn't be more readable here. What I mean by that is create a companion object that will have methods like `createWithMigrations`, `createWithSchema` and `createIfCompatible` or something like that instead of secondary constructors. WDYT?

Reply:
I don't think that factory is a good idea. Kotlin has very nice system of named arguments so when someone need to create database with migrations he can just use 
```kotlin
DatabaseDriver(context, dbName, migrations = MigrationSet())
```
with Schema
```kotlin
DatabaseDriver(context, dbName, schema = Schema())
```
and to check if compatible 
```kotlin
DatabaseDriver(context, dbName, schemaVersion = SchemaVersion())
```

If you are using named parameters when creating DatabaseDriver there is no confusion about which kind will be created